src/main/controllers/models/tx-history.js

```diff
@@ -0,0 +1,79 @@
+const Promise = require('bluebird');
+const isSyncing = require('../tx-history-service').isSyncing;
+
+module.exports = function (app, sqlLiteService) {
+    const TABLE_NAME = 'tx_history';
+    const Controller = function () { };
+
+    let knex = sqlLiteService.knex;
+
+    /**
+     *
+     */
+    Controller.addOrUpdate = _addOrUpdate;
+    Controller.findByTxHash = _findByTxHash;
+    Controller.findByPublicKey = _findByPublicKey;
+    Controller.findByPublicKeyAndTokenSymbol = _findByPublicKeyAndTokenSymbol;
+    Controller.findByPublicKeyAndContractAddress = _findByPublicKeyAndContractAddress;
+
+    async function _addOrUpdate(txHistory) {
+        let records = await _findByTxHash(txHistory.hash);
+        let record = records ? records[0] : null;
+        if (record) {
+            Object.assign(record, txHistory);
+            return sqlLiteService.updateById(TABLE_NAME, record);
+        }
+        return sqlLiteService.insertIntoTable(TABLE_NAME, txHistory);
+    }
+
+    async function _findByTxHash(hash) {
+        return await knex(TABLE_NAME).where({ hash: hash });
+    }
+
+    async function _findByPublicKeyAndContractAddress(publicKey, contractAddress) {
+        publicKey = publicKey.toLowerCase();
+        return new Promise((resolve, reject) => {
```

Comment:
`knex` returns promises by default. We don't need to wrap them in this external promise.

Reply:
@dfcowell the reason of using external promise is, that I'm returning `isSyncing: isSyncing(publicKey` additionally

Reply:
```
return knex(TABLE_NAME).where({ from: publicKey, contractAddress })
  .orWhere({ to: publicKey, contractAddress }).orderBy('timeStamp', 'desc')
  .then(rows => ({
    data: rows,
    isSyncing: isSyncing(publicKey)
  }))